src/task.rs

```diff
@@ -319,6 +335,239 @@ impl<'a> Drop for CriticalSectionGuard<'a> {
+                );
+                #[cfg(not(esp_idf_esp_task_wdt))]
+                let duration = core::time::Duration::from_secs(5);
+                Self {
+                    duration,
+                    panic_on_trigger: cfg!(esp_idf_esp_task_wdt_panic),
+                    subscribed_idle_tasks: {
+                        let mut subscribed_idle_tasks = enumset::EnumSet::empty();
+                        if cfg!(esp_idf_esp_task_wdt_check_idle_task_cpu0) {
+                            subscribed_idle_tasks |= Core::Core0;
+                        }
+                        #[cfg(any(esp32, esp32s3))]
+                        if cfg!(esp_idf_esp_task_wdt_check_idle_task_cpu1) {
+                            subscribed_idle_tasks |= Core::Core1;
+                        }
+                        subscribed_idle_tasks
+                    },
+                }
+            }
+        }
+
+        impl Default for Config {
+            fn default() -> Self {
+                Self::new()
+            }
+        }
+
+        #[cfg(not(esp_idf_version_major = "4"))]
+        impl From<Config> for esp_task_wdt_config_t {
+            fn from(config: Config) -> Self {
+                esp_task_wdt_config_t {
+                    timeout_ms: config.duration.as_millis() as u32,
+                    trigger_panic: config.panic_on_trigger,
+                    idle_core_mask: config.subscribed_idle_tasks.as_u32(),
+                }
+            }
+        }
+    }
+
+    pub struct TWDTDriver<'d> {
+        init_by_idf: bool,
+        _marker: PhantomData<&'d mut ()>,
+    }
+
+    static TWDT_DRIVER_REF_COUNT: AtomicUsize = AtomicUsize::new(0);
+
+    impl<'d> TWDTDriver<'d> {
+        pub fn new(
+            _twdt: impl Peripheral<P = TWDT> + 'd,
+            config: &config::Config,
+        ) -> Result<Self, EspError> {
+            TWDT_DRIVER_REF_COUNT.fetch_add(1, Ordering::SeqCst);
+            let init_by_idf = Self::watchdog_is_init_by_idf();
+
+            #[cfg(not(esp_idf_version_major = "4"))]
+            if !init_by_idf {
+                esp!(unsafe { esp_task_wdt_init(config.into()) })?;
+            } else {
+                esp!(unsafe { esp_task_wdt_reconfigure(config.into()) })?;
+            }
+
+            #[cfg(esp_idf_version_major = "4")]
+            esp!(unsafe {
+                esp_task_wdt_init(config.duration.as_secs() as u32, config.panic_on_trigger)
+            })?;
+
+            #[cfg(esp_idf_version_major = "4")]
+            Self::subscribe_idle_tasks(config.subscribed_idle_tasks);
+
+            Ok(Self {
+                init_by_idf,
+                _marker: Default::default(),
+            })
+        }
+
+        pub fn watch_current_task(&mut self) -> Result<WatchdogSubscription<'_>, EspError> {
+            esp!(unsafe { esp_task_wdt_add(core::ptr::null_mut()) })?;
+            Ok(WatchdogSubscription::new())
+        }
+
+        fn subscribe_idle_tasks(cores: enumset::EnumSet<Core>) {
```

Comment:
Should return `Result<(), EspError>`. Let's not swallow ESP IDF failures, unless really necessary.